src/Rings/msolve/msolve.jl

```diff
@@ -0,0 +1,205 @@
+    lens, cfs, exps   = convert_singular_ideal_to_array(J)
+
+    res_ld    = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
+    res_dim   = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
+    res_dquot = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
+    res_len   = ccall(:malloc, Ptr{Ptr{Cint}}, (Csize_t, ), sizeof(Ptr{Cint}))
+    res_cf    = ccall(:malloc, Ptr{Ptr{Cvoid}}, (Csize_t, ), sizeof(Ptr{Cvoid}))
+    nb_sols   = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
+    sols_num  = ccall(:malloc, Ptr{Ptr{Cvoid}}, (Csize_t, ), sizeof(Ptr{Cvoid}))
+    sols_den  = ccall(:malloc, Ptr{Ptr{Cint}}, (Csize_t, ), sizeof(Ptr{Cint}))
+    ccall((:msolve_julia, libmsolve), Cvoid,
+        (Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Ptr{Cint}}, Ptr{Cvoid}, Ptr{Cint},
+         Ptr{Cvoid}, Ptr{Ptr{Cint}}, Ptr{Cint}, Ptr{Cint}, Ptr{Cvoid},
+         Ptr{Ptr{Cchar}}, Ptr{Cchar}, Int, Int, Int, Int,
+         Int, Int, Int, Int, Int, Int, Int, Int, Int, Int),
+        res_ld, res_dim, res_dquot, res_len, res_cf, nb_sols, sols_num, sols_den, lens,
+        exps, cfs, variable_names, "/dev/null", field_char, mon_order, nr_vars,
+        nr_gens, initial_hts, nr_thrds, max_nr_pairs, reset_ht, la_option,
+        print_gb, get_param, genericity_handling, precision, info_level)
+    # convert to julia array, also give memory management to julia
+    jl_ld       = unsafe_load(res_ld)
+
+    jl_dim      = unsafe_load(res_dim)
+    jl_dquot    = unsafe_load(res_dquot)
+
+    jl_nb_sols  = unsafe_load(nb_sols)
+    jl_len      = Base.unsafe_wrap(Array, unsafe_load(res_len), jl_ld)
+
+    nterms  = 0
+    
+    # set dimension
+    I.dim = jl_dim
+    if jl_dim > 0
+        @info "Dimension is greater than zero, no solutions provided."
+        return []
+    end
+    if jl_nb_sols == 0
+        @info "The system has no solution."
+        return []
+    end
+
+    [nterms += jl_len[i] for i=1:jl_ld]
+    # if 0 == field_char
+        res_cf_conv   = unsafe_load(res_cf)
+        jl_cf         = reinterpret(Ptr{BigInt}, res_cf_conv)
+        sols_num_conv = unsafe_load(sols_num)
+        jl_sols_num   = reinterpret(Ptr{BigInt}, sols_num_conv)
+        sols_den_conv = unsafe_load(sols_den)
+        jl_sols_den   = reinterpret(Ptr{Int32}, sols_den_conv)
+    # elseif isprime(field_char)
+    #     res_cf_conv = unsafe_load(res_cf)
+    #     jl_cf       = reinterpret(Ptr{Int32}, res_cf_conv)
+    #     sols_num_conv = unsafe_load(sols_num)
+    #     jl_sols_num   = reinterpret(Ptr{Int32}, sols_num_conv)
+    # end
+
+    for i in 1:jl_nb_sols*nr_vars
+        tmpcf = BigInt(unsafe_load(jl_sols_num, i))
+    end
+    solutions = Array{Array{fmpq, 1}, 1}(undef, jl_nb_sols)
+    for i in 1:jl_nb_sols
+        tmp = Array{Nemo.fmpq, 1}(undef, nr_vars)
+        for j in 1:nr_vars
+            tmp[j]  = BigInt(unsafe_load(jl_sols_num, (i-1)*nr_vars+j)) // BigInt(2)^unsafe_load(jl_sols_den, (i-1)*nr_vars+j)
```

Comment:
Since you want to produce `fmpq` I think it makes more sense to use `fmpz` for the intermediates instead of `BigInt` ? 
```suggestion
            tmp[j]  = fmpz(unsafe_load(jl_sols_num, (i-1)*nr_vars+j)) // fmpz(2)^unsafe_load(jl_sols_den, (i-1)*nr_vars+j)
```
Moreover, it seems you only divide by powers of two, so this should also do the same job but more efficiently:
```suggestion
            tmp[j]  = fmpq(unsafe_load(jl_sols_num, (i-1)*nr_vars+j)) >> unsafe_load(jl_sols_den, (i-1)*nr_vars+j)
```